test/e2e/util.go

```diff
@@ -889,6 +911,142 @@ func waitReplicaSetReady(ctx *testContext, name string) error {
+
+	used := map[string]*schedulerapi.Resource{}
+
+	for _, pod := range pods.Items {
+		nodeName := pod.Spec.NodeName
+		if len(nodeName) == 0 || pod.DeletionTimestamp != nil {
+			continue
+		}
+
+		if pod.Status.Phase == v1.PodSucceeded || pod.Status.Phase == v1.PodFailed {
+			continue
+		}
+
+		if _, found := used[nodeName]; !found {
+			used[nodeName] = schedulerapi.EmptyResource()
+		}
+
+		for _, c := range pod.Spec.Containers {
+			resource := schedulerapi.NewResource(c.Resources.Requests)
+			used[nodeName].Add(resource)
+		}
+	}
+
+	//var minCPU, minMemory
+	minCPU := req.Cpu()
+	minMemory := req.Memory()
+	resourceRichNode := 0
+
+	//init placeholders
+	placeHolders := map[string]v1.ResourceList{}
+
+	for _, node := range nodes.Items {
+		if len(node.Spec.Taints) != 0 {
+			continue
+		}
+		minCPUMilli := float64(minCPU.MilliValue())
+		minMemoryValue := float64(minMemory.Value())
+		currentAllocatable := schedulerapi.NewResource(node.Status.Allocatable)
+
+		if res, found := used[node.Name]; found {
+			currentAllocatable.Sub(res)
+		}
+
+		phCPU := currentAllocatable.MilliCPU
+		phMemory := currentAllocatable.Memory
+
+		if minCPUMilli <= currentAllocatable.MilliCPU && minMemoryValue <= currentAllocatable.Memory {
+			resourceRichNode = resourceRichNode + 1
+			if resourceRichNode <= reqNum {
+				phCPU = currentAllocatable.MilliCPU - minCPUMilli
+				phMemory = currentAllocatable.Memory - minMemoryValue
+			}
+		}
+
+		phCPUQuantity := resource.NewMilliQuantity(int64(phCPU), resource.BinarySI)
+		phMemoryQuantity := resource.NewQuantity(int64(phMemory), resource.BinarySI)
+		placeHolders[node.Name] = v1.ResourceList{"cpu": *phCPUQuantity, "memory": *phMemoryQuantity}
+	}
+
+	if resourceRichNode < reqNum {
+		return false, lagencyerror.New("Failed to setup environment, you need to have at least " + strconv.Itoa(len(req)) + " worker node.")
+	}
+
+	for nodeName, res := range placeHolders {
+		err := createPlaceHolder(ctx, res, nodeName)
+		Expect(err).NotTo(HaveOccurred())
+	}
+
+	return true, nil
+}
+
+func createPlaceHolder(ctx *testContext, phr v1.ResourceList, nodeName string) error {
+	pod := &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      nodeName + "-placeholder",
+			Namespace: "default",
+			Labels: map[string]string{
+				"role": "placeholder",
+			},
+		},
+		Spec: corev1.PodSpec{
+			Containers: []corev1.Container{
+				{
+					Name: "placeholder",
+					Resources: corev1.ResourceRequirements{
+						Requests: phr,
+						Limits:   phr,
+					},
+					Image: defaultNginxImage,
+				},
+			},
+			NodeName: nodeName,
+		},
+	}
+	_, err := ctx.kubeclient.CoreV1().Pods("default").Create(context.TODO(), pod, metav1.CreateOptions{})
+	return err
+}
+
+func deletePlaceHolder(ctx *testContext) {
+	podList, err := ctx.kubeclient.CoreV1().Pods("default").List(context.TODO(), metav1.ListOptions{})
+	Expect(err).NotTo(HaveOccurred())
+	for _, pod := range podList.Items {
+		if pod.Labels["role"] == "placeholder" {
```

Comment:
This is not efficient loop all the pods, you can use field LabelSelector instead

Reply:
OK